servicetalk-http-api/src/main/java/io/servicetalk/http/api/StreamingHttpService.java

```diff
@@ -46,4 +52,94 @@
     default Completable closeAsync() {
         return completed();
     }
+
+    /**
+     * Wraps the passed {@link StreamingHttpService} to apply the provided {@link HttpExecutionStrategy} for offloading.
+     *
+     * @param strategy {@link HttpExecutionStrategy} to use for offloading.
+     * @param executor {@link Executor} to use as executor or {@code null} to use Execution context executor.
+     * @param shouldOffload If {@link BooleanSupplier} returns {@code true} then offload to executor otherwise continue
+     * execution on calling thread.
+     * @param service {@link StreamingHttpService} to wrap.
+     * @return Wrapped {@link StreamingHttpService}.
+     */
+    static StreamingHttpService offloadService(final HttpExecutionStrategy strategy,
+                                               @Nullable final Executor executor, final BooleanSupplier shouldOffload,
+                                               final StreamingHttpService service) {
+        return strategy.hasOffloads() ?
+                new StreamingHttpService() {
+                    @Override
+                    public Single<StreamingHttpResponse> handle(final HttpServiceContext ctx,
+                                                                StreamingHttpRequest request,
+                                                                final StreamingHttpResponseFactory responseFactory) {
+                        // We compute the difference between the ExecutionStrategy from the current ExecutionContext and
+                        // this ExecutionStrategy to understand if we need to offload more than we already offloaded:
+                        final HttpExecutionStrategy diff =
+                                difference(ctx.executionContext().executionStrategy(), strategy);
+
+                        Executor useExecutor = null != executor ? executor : ctx.executionContext().executor();
+
+                        // The service should see this ExecutionStrategy inside the ExecutionContext:
+                        final HttpServiceContext wrappedCtx =
```

Comment:
ideally we can avoid allocation/wrapping on each request. I guess if we need to we could cache this object in the future.

Reply:
I was pretty sure the context never changed, but it could for conditional filters, so I didn't implement caching. If we think it is worthwhile I can add it.